modules/core/src/main/scala/org/scalasteward/core/buildtool/BuildToolDispatcher.scala

```diff
@@ -25,44 +25,73 @@ import org.scalasteward.core.data.{Resolver, Scope}
 import org.scalasteward.core.scalafix.Migration
 import org.scalasteward.core.scalafmt.ScalafmtAlg
 import org.scalasteward.core.util.Nel
-import org.scalasteward.core.vcs.data.Repo
+import org.scalasteward.core.vcs.data.{BuildRoot, Repo}
+import org.scalasteward.core.repoconfig.RepoConfigAlg
 
-trait BuildToolDispatcher[F[_]] extends BuildToolAlg[F]
+trait BuildToolDispatcher[F[_]] extends BuildToolAlg[F] {
+  def runMigrationsForAllBuildRoots(repo: Repo, migrations: Nel[Migration]): F[Unit]
+  def getDependenciesForAllBuildRoots(repo: Repo): F[List[Scope.Dependencies]]
+}
```

Comment:
I find it unfortunate that `BuildToolDispatcher` has now methods that work with `Repo` and `BuildRoot` but the only methods used outside of `BuildToolDispatcher` are the ones with `Repo`. Have you tried to add a second type parameter to `BuildToolAlg` as mentioned in [my comment](https://github.com/scala-steward-org/scala-steward/pull/1093#issuecomment-692246532):

> BuildToolDispatcher and the build tool traits like SbtAlg inherit from BuildToolAlg which uses Repo for its methods. Since SbtAlg, MavenAlg, and MillAlg should now operate on BuildRoots instead of Repo but BuildToolDispatcher should still use Repo we need to add a new type parameter to the BuildToolAlg, i.e. BuildToolAlg[F, R] where R would be either a Repo or a BuildRoot.

?

I think with this change, the above additions aren't necessary and the `BuildToolDispatcher`'s callsites (in `EditAlg` and `RepoCacheAlg`) also do not need to be changed.

Reply:
Right that might work, I totally didn't see that part of the comment somehow. I'll change this asap.

Reply:
Switched over, yeah that is a much nicer approach. Technically it should be possible to have contravariant typeclasses on `BuildToolAlg` that would allow the dispatcher to work easily both for `Repo` and `BuildRoot` I didn't do that since I'm a bit pressed for time now but if you see any value in that I'll happily change it to that.

Reply:
Looks good to me now, thanks. I think there is no need for more changes since the dispatcher is only used with `Repo` for now.